### PR TITLE
Add Power BI Data Gateway VM Quickstart

### PR DESCRIPTION
**Pull request by daltskin**
PR to install and configure a [Power BI Data Gateway](https://docs.microsoft.com/en-us/data-integration/gateway/) within an Azure - please see the README.md for more details.

# PR Checklist

Check these items before submitting a PR... 

[Contribution Guide](https://github.com/Azure/azure-quickstart-templates/blob/master/1-CONTRIBUTION-GUIDE/README.md)

[Best Practice Guide](https://github.com/Azure/azure-quickstart-templates/blob/master/1-CONTRIBUTION-GUIDE/best-practices.md)

- [x ] - Please check this box once you've submitted the PR if you've read through the Contribution Guide and best practices checklist.

## Changelog

*
*
*

### CODE REVIEW

**Comment by daltskin** [comment]
Not sure how to test this as pre-requires AAD App to be setup (as per the README.md)

File: 101-vm-with-powerbi-data-gateway/azuredeploy.json [hunk 1]
```diff
@@ -0,0 +1,301 @@
...
+      "metadata": {
+        "description": "Admin password on VM"
+      }
+    },
+    "vmSize": {
+      "type": "string",
+      "metadata": {
+        "description": "Desired Size of the VM. Any valid option accepted but if you choose premium storage type you must choose a DS class VM size."
+      },
+      "defaultValue": "Standard_DS2"
```
Comment (bmoore-msft) [inline 101-vm-with-powerbi-data-gateway/azuredeploy.json:40]: move this above metadata

File: 101-vm-with-powerbi-data-gateway/azuredeploy.json [hunk 2]
```diff
@@ -0,0 +1,301 @@
...
+      }
+    },
+    "aadAppId": {
+      "type": "string",
+      "metadata": {
+        "description": "AAD Application Id"
+      }
+    },
+    "aadAppSecret": {
+      "type": "string",
```
Comment (bmoore-msft) [inline 101-vm-with-powerbi-data-gateway/azuredeploy.json:63]: secure string

File: 101-vm-with-powerbi-data-gateway/azuredeploy.json [hunk 3]
```diff
@@ -0,0 +1,301 @@
...
+    "tenantId": {
+      "type": "string",
+      "defaultValue": "[subscription().tenantId]",
+      "metadata": {
+        "description": "AAD Tenant Id"
+      }
+    },
+    "gatewayName": {
+      "type": "string",
+      "metadata": {
```
Comment (bmoore-msft) [inline 101-vm-with-powerbi-data-gateway/azuredeploy.json:77]: add a defaultValue using uniqueString

File: 101-vm-with-powerbi-data-gateway/azuredeploy.json [hunk 4]
```diff
@@ -0,0 +1,301 @@
...
+        "description": "Power BI Data Gateway name must be unique in subscription"
+      }
+    },
+    "gatewayRegionKey": {
+      "type": "string",
+      "defaultValue": "[resourceGroup().location]",
+      "metadata": {
+        "description": "Power BI Data Gateway RegionKey eg: northeurope"
```
Comment (bmoore-msft) [inline 101-vm-with-powerbi-data-gateway/azuredeploy.json:85]: how is this different than location?

File: 101-vm-with-powerbi-data-gateway/azuredeploy.json [hunk 5]
```diff
@@ -0,0 +1,301 @@
...
+      "metadata": {
+        "description": "Power BI Data Gateway Administrator Ids (comma separted list of AAD Object Ids)"
+      }
+    },
+    "_artifactsLocation": {
+      "type": "string",
+      "metadata": {
+        "description": "The base URI where artifacts required by this template are located."
+      },
+      "defaultValue": "https://raw.githubusercontent.com/Azure/azure-quickstart-templates/master/101-vm-with-powerbi-data-gateway/"
```
Comment (bmoore-msft) [inline 101-vm-with-powerbi-data-gateway/azuredeploy.json:105]: move this above metadata - won't comment on the rest

File: 101-vm-with-powerbi-data-gateway/azuredeploy.json [hunk 6]
```diff
@@ -0,0 +1,301 @@
...
+    }
+  },
+  "variables": {
+    "storageName": "[concat(uniquestring(resourceGroup().id), 'standardsa')]",
+    "vnetName": "VNet01",
+    "vnet01Prefix": "10.0.0.0/16",
+    "vnet01Subnet1Prefix": "10.0.0.0/24",
+    "vmImagePublisher": "MicrosoftWindowsServer",
+    "vmImageOffer": "WindowsServer",
+    "vmSKU": "2016-Datacenter",
```
Comment (bmoore-msft) [inline 101-vm-with-powerbi-data-gateway/azuredeploy.json:122]: any reason we can't use 2019?

File: 101-vm-with-powerbi-data-gateway/images/gatewayinstall.json [hunk 7]
```diff
@@ -0,0 +1,14 @@
+[
```
Comment (bmoore-msft) [inline 101-vm-with-powerbi-data-gateway/images/gatewayinstall.json:1]: is this file needed?

**Review by bmoore-msft — CHANGES_REQUESTED** [review]
If you can use a dummy app registration for deployment the CI engine has one you can use... if you need specific perms then we'll need to run a manual test.  In that case, set the validation type in metadata.json and provide a PowerShell script that will configure the prereqs you need and outputs the param values that will be passed to the test.

File: 101-vm-with-powerbi-data-gateway/azuredeploy.json (hunk 3 above)
Comment (daltskin) [reply 101-vm-with-powerbi-data-gateway/azuredeploy.json:78]: done

File: 101-vm-with-powerbi-data-gateway/azuredeploy.json (hunk 4 above)
Comment (daltskin) [reply 101-vm-with-powerbi-data-gateway/azuredeploy.json:86]: PBI Data Gateway region can be set independently from the registered services tenant - see: https://docs.microsoft.com/en-us/data-integration/gateway/service-gateway-data-region

File: 101-vm-with-powerbi-data-gateway/azuredeploy.json (hunk 5 above)
Comment (daltskin) [reply 101-vm-with-powerbi-data-gateway/azuredeploy.json]: Done

File: 101-vm-with-powerbi-data-gateway/azuredeploy.json (hunk 6 above)
Comment (daltskin) [reply 101-vm-with-powerbi-data-gateway/azuredeploy.json]: Done

File: 101-vm-with-powerbi-data-gateway/azuredeploy.json (hunk 2 above)
Comment (daltskin) [reply 101-vm-with-powerbi-data-gateway/azuredeploy.json]: Done

**Comment by daltskin** [comment]
Hi @bmoore-msft  - when you say "use a dummy app registration" - do you mean, set aadAppId within params.json to GEN-GUID?

**Comment by daltskin** [comment]
ok, starting to figure this out.  I've added a prereqs script with a deployscripts to create the AAD app - but not sure why it's not seeing any outputs though - this looks ok to me following this: https://docs.microsoft.com/en-us/azure/azure-resource-manager/templates/deployment-script-template?tabs=CLI#work-with-outputs-from-cli-script

**Comment by daltskin** [comment]
@bmoore-msft does the CI account have permissions to register a new AAD App?  If not, how can I try the dummy app creds?

**Comment by bmoore-msft** [comment]
@daltskin - I updated the param file to use a SP we have for CI, this SP doesn't have any permissions on the sub (which looks to be similar to your deploymentScript) but the deployment still fails.  You can take a look at the pipeline results to see if anything jumps out, though sounds like we might need more logging to know where the script is failing.

**Comment by daltskin** [comment]
Thanks @bmoore-msft - all of the logging is reported within the CustomScriptExtension itself, but as the CI build tidies up the resources after itself, I'm unable to see the Detailed Status (as per: https://github.com/Azure/azure-quickstart-templates/blob/ce42f38e7c00c71681c5305c0212fa9d1d8e6a7b/101-vm-with-powerbi-data-gateway/README.md#troubleshooting).  

Looking at the CLI/PowerShell custom extension [documentation](https://docs.microsoft.com/en-us/azure/azure-arc/servers/manage-vm-extensions-cli), it doesn't look like you can extract the Detailed Status either to persist the output within the CI process.  Are able to run the CI (without the clean up tasks) and extract the Detailed Status from the VM's custom extension manually so I can take a look?  (Or give me read permissions to the resource group so I can help troubleshoot)?

**Comment by daltskin** [comment]
@bmoore-msft is the CI Azure subscription linked to an M365 subscription?

**Comment by daltskin** [comment]
@bmoore-msft it looks like there is a way to extract the `Detailed Status` data from the VM extension by issuing the following command before the clean-up process within the CI process:

```bash
az rest  --url /subscriptions/{subscriptionId}/resourceGroups/{rgName}/providers/Microsoft.Compute/virtualMachines/{vmName}/instanceView?api-version=2020-06-01 -o json
```

So in the CI build example is:

```bash
az rest  --url /subscriptions/0cec7090-2e08-4498-9337-eb96ade50821/resourceGroups/azdo-666c8674-05ca-465a-9669-d9a2b2a3599c/providers/Microsoft.Compute/virtualMachines/PBIGateway/instanceView?api-version=2020-06-01 -o json
```
This would need to be run whilst the VM exists, so before the "Clean Up ResourceGroup (main)" task

**Comment by bmoore-msft** [comment]
I'm not sure what you mean by the link to the M365 subscription so my guess is "no" on that piece.

One thing I noticed in the pipeline logs is that the app/secret are not being output, they are empty:

https://dev.azure.com/azurequickstarts/azure-quickstart-templates/_build/results?buildId=43270&view=logs&j=275f1d19-1bd8-5591-b06b-07d489ea915a&t=8ebc9149-1226-5c43-6fb4-06b80a585d5d&l=71

So that may be the first issue.  Also noticed that you're not assigning any perms, if none are needed you should be able to use the existing app registration for the pipeline.  I'll try that really quick to see if it unblocks, feel free to undo the change as appropriate.

**Comment by bmoore-msft** [comment]
nm on that last bit - I see that's already been tried... let me run a manual deployment and see what we get.

**Comment by bmoore-msft** [comment]
Here's the output from the vm:

```
               "extensions": [
                 {
                   "name": "PBIGateway",
                   "type": "Microsoft.Compute.CustomScriptExtension",
                   "typeHandlerVersion": "1.10.9",
                   "substatuses": [
                     {
                       "code": "ComponentStatus/StdOut/succeeded",
                       "level": "Info",
                       "displayStatus": "Provisioning succeeded",
                       "message": "Installing PowerShell v7\r\nDownload & install PowerShell v7 from script\r\nVERBOSE: About to download package from
             \r\n'https://github.com/PowerShell/PowerShell/releases/download/v7.1.1/PowerShell-7.1.1-win-x64.msi'\r\nInstalling PBI Gateway under PowerShell v7 shell\r\nResult=Installing DataGateway PS Module\r\nConnect to the Data Gateway
             Service\r\nWARNING: (False) MSAL 4.7.0.0 MSAL.NetCore Microsoft Windows 10.0.17763 [02/04/2021 15:32:47 - 401ebc0b-37af-4560-8a90-404dc11c8873] (UnknownClient: 0.0.0.0) Exception type: Microsoft.Identity.Client.MsalServiceException\r\n,
             ErrorCode: invalid_client\r\nHTTP StatusCode 401\r\nCorrelationId 401ebc0b-37af-4560-8a90-404dc11c8873\r\n\r\nWARNING: (False) MSAL 4.7.0.0 MSAL.NetCore Microsoft Windows 10.0.17763 [02/04/2021 15:32:47 -
             401ebc0b-37af-4560-8a90-404dc11c8873] (UnknownClient: 0.0.0.0) Fetching a new AT failed. Is AAD down? False. Is there an AT in the cache that is usable? False\r\nWARNING: (False) MSAL 4.7.0.0 MSAL.NetCore Microsoft Windows 10.0.17763
             [02/04/2021 15:32:47 - 401ebc0b-37af-4560-8a90-404dc11c8873] (UnknownClient: 0.0.0.0) Either the exception does not indicate a problem with AAD or the token cache does not have an AT that is usable.\r\nThe wrong application (public or
             confidential) is being used with this authentication flow. Potential issue: the redirect URI is valid, but it has been configured for the wrong app type. Check the configuration of the app being used in the app registration portal. See
             https://aka.ms/msal-net-invalid-client for details.  Original exception: AADSTS7000222: The provided client secret keys are expired. Visit the Azure Portal to create new keys for your app, or consider using certificate credentials for
             added security: https://docs.microsoft.com/azure/active-directory/develop/active-directory-certificate-credentials\r\nTrace ID: 09a24f10-3d9c-449a-abc8-7e030504f800\r\nCorrelation ID: 401ebc0b-37af-4560-8a90-404dc11c8873\r\nTimestamp:
             2021-02-04 15:32:47Z\r\n \\n\\rExitCode=1\r\n"
                     },
                     {
                       "code": "ComponentStatus/StdErr/succeeded",
                       "level": "Info",
                       "displayStatus": "Provisioning succeeded",
                       "message": "Invoke-Process : Failed to run process: exitCode=1, errVariable=, errContent=\u001b[91mConnect-DataGatewayServiceAccount:
             \r\n\u001b[0mC:\\Packages\\Plugins\\Microsoft.Compute.CustomScriptExtension\\1.10.9\\Downloads\\0\\pbiGatewayInstall.ps1:54\r\n\u001b[96mLine |\r\n\u001b[96m  54 | \u001b[0m . onnected = (\u001b[96mConnect-DataGatewayServiceAccount
             -ApplicationId $AppId -\u001b[0m .\r\n\u001b[96m     | \u001b[91m               ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\r\n\u001b[91m\u001b[96m     | \u001b[91mThe wrong application (public or confidential) is being
             used with this authentication flow. Potential\r\n\u001b[96m     | \u001b[91missue: the redirect URI is valid, but it has been configured for the wrong app type. Check the\r\n\u001b[96m     | \u001b[91mconfiguration of the app being used
             in the app registration portal. See\r\n\u001b[96m     | \u001b[91mhttps://aka.ms/msal-net-invalid-client for details.  Original exception: AADSTS7000222: The provided\r\n\u001b[96m     | \u001b[91mclient secret keys are expired. Visit
             the Azure Portal to create new keys for your app, or consider\r\n\u001b[96m     | \u001b[91musing certificate credentials for added security:\r\n\u001b[96m     |
             \u001b[91mhttps://docs.microsoft.com/azure/active-directory/develop/active-directory-certificate-credentials\r\n\u001b[96m     | \u001b[91mTrace ID: 09a24f10-3d9c-449a-abc8-7e030504f800 Correlation ID:
             401ebc0b-37af-4560-8a90-404dc11c8873\r\n\u001b[96m     | \u001b[91mTimestamp: 2021-02-04 15:32:47Z\r\n\u001b[0m\r\n\u001b[91mWrite-Error: \u001b[91mError: Connecting to Data Gateway Service\u001b[0m\r\n, outContent=Installing
             DataGateway PS Module\r\nConnect to the Data Gateway Service\r\nWARNING: (False) MSAL 4.7.0.0 MSAL.NetCore Microsoft Windows 10.0.17763 [02/04/2021 15:32:47 - \r\n401ebc0b-37af-4560-8a90-404dc11c8873] (UnknownClient: 0.0.0.0) Exception
             type: \r\nMicrosoft.Identity.Client.MsalServiceException\r\n, ErrorCode: invalid_client\r\nHTTP StatusCode 401\r\nCorrelationId 401ebc0b-37af-4560-8a90-404dc11c8873\r\nWARNING: (False) MSAL 4.7.0.0 MSAL.NetCore Microsoft Windows
             10.0.17763 [02/04/2021 15:32:47 - \r\n401ebc0b-37af-4560-8a90-404dc11c8873] (UnknownClient: 0.0.0.0) Fetching a new AT failed. Is AAD down? False. Is there \r\nan AT in the cache that is usable? False\r\nWARNING: (False) MSAL 4.7.0.0
             MSAL.NetCore Microsoft Windows 10.0.17763 [02/04/2021 15:32:47 - \r\n401ebc0b-37af-4560-8a90-404dc11c8873] (UnknownClient: 0.0.0.0) Either the exception does not indicate a problem with \r\nAAD or the token cache does not have an AT
             that is usable.\r\nThe wrong application (public or confidential) is being used with this authentication flow. Potential issue: the \r\nredirect URI is valid, but it has been configured for the wrong app type. Check the configuration of
             the app being \r\nused in the app registration portal. See https://aka.ms/msal-net-invalid-client for details.  Original exception: \r\nAADSTS7000222: The provided client secret keys are expired. Visit the Azure Portal to create new
             keys for your app, or \r\nconsider using certificate credentials for added security: \r\nhttps://docs.microsoft.com/azure/active-directory/develop/active-directory-certificate-credentials\r\nTrace ID:
             09a24f10-3d9c-449a-abc8-7e030504f800\r\nCorrelation ID: 401ebc0b-37af-4560-8a90-404dc11c8873\r\nTimestamp: 2021-02-04 15:32:47Z\r\n.\r\nAt
             C:\\Packages\\Plugins\\Microsoft.Compute.CustomScriptExtension\\1.10.9\\Downloads\\0\\setup.ps1:61 char:11\r\n+ $result = Invoke-Process \"$env:ProgramFiles\\PowerShell\\7\\pwsh.exe\" $p ...\r\n+
             ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\r\n    + CategoryInfo          : NotSpecified: (:) [Write-Error], WriteErrorException\r\n    + FullyQualifiedErrorId :
             Microsoft.PowerShell.Commands.WriteErrorException,Invoke-Process\r\n \r\nError within pbiGateway.ps1: \r\nAt C:\\Packages\\Plugins\\Microsoft.Compute.CustomScriptExtension\\1.10.9\\Downloads\\0\\setup.ps1:70 char:2\r\n+     throw
             $progressMsg\r\n+     ~~~~~~~~~~~~~~~~~~\r\n    + CategoryInfo          : OperationStopped: (Error within pbiGateway.ps1: :String) [], RuntimeException\r\n    + FullyQualifiedErrorId : Error within pbiGateway.ps1: \r\n \r\n"
                     }
                   ],
                   "statuses": [
                     {
                       "code": "ProvisioningState/failed/1",
                       "level": "Error",
                       "displayStatus": "Provisioning failed",
                       "message": "Command execution finished, but failed because it returned a non-zero exit code of: '1'. The command had an error output of: 'Invoke-Process : Failed to run process: exitCode=1, errVariable=,
             errContent=\u001b[91mConnect-DataGatewayServiceAccount: \r\n\u001b[0mC:\\Packages\\Plugins\\Microsoft.Compute.CustomScriptExtension\\1.10.9\\Downloads\\0\\pbiGatewayInstall.ps1:54\r\n\u001b[96mLine |\r\n\u001b[96m  54 | \u001b[0...' For
             more information, check the instance view by executing Get-AzVmssVm or Get-AzVm (https://aka.ms/GetAzVm). These commands can be executed using CloudShell (https://aka.ms/CloudShell)"
                     }
                   ]
                 }
               ],

```

**Comment by daltskin** [comment]
Thanks for the log.  Looks like the SP being used to access the [Data Gateway service](https://docs.microsoft.com/en-us/powershell/module/datagateway.profile/connect-datagatewayserviceaccount?view=datagateway-ps) is not valid.

If the Azure subscription in use for CI doesn't have a license for Office 365, I'm don't think this will work :(

image

**Comment by bmoore-msft** [comment]
Here's what we have to the test sub:

image

So that seems ok - the SP is still active:

image

but it doesn't have any permissions (which is the same as the one we were creating in the deploymentScript).  What perms does the SP need?

**Comment by daltskin** [comment]
Should just need:
```powershell
# In the Azure portal these will show as the Permission names "Tenant.Read.All" and "Tenant.ReadWrite.All" with the "Application" type under the Power BI Service API
$resourceAccessItems = @(
                        @{Id="654b31ae-d941-4e22-8798-7add8fdf049f";Type="Role"},
                        @{Id="28379fa9-8596-4fd9-869e-cb60a93b5d84";Type="Role"}
                        );
# In the Azure portal the ResourceAppId below is the Power BI Service
$resourceAccess = @{ResourceAppId="00000009-0000-0000-c000-000000000000"; ResourceAccess=$resourceAccessItems};
```

**Comment by daltskin** [comment]
> One thing I noticed in the pipeline logs is that the app/secret are not being output, they are empty:
> 
> https://dev.azure.com/azurequickstarts/azure-quickstart-templates/_build/results?buildId=43270&view=logs&j=275f1d19-1bd8-5591-b06b-07d489ea915a&t=8ebc9149-1226-5c43-6fb4-06b80a585d5d&l=71
 
@bmoore-msft I've been troubleshooting this to see what's happening.  Adding additional error logging locally I can see `Insufficient privileges to complete the operation` coming back from running the following command inside the DeploymentScripts - ScriptContent `az ad sp create-for-rbac -o json --name $appName`

So the Managed Identity used with the prereq.azuredeploy.json doesn't have the correct permissions to create the SPN, which is later needed to login to the O365 Power BI Tenant.  However, I've tried changing the MIS to both User Access Administrator and Owner but still get the insufficient privileges error.  Do you know if this is possible?  Or is this the wrong approach?

image

Alternatively, using the SPN you already have for the CI, with it's appId & password in the output for the prereqs should allow the complete deployment to succeed.  Since I tested this on a new SPN without any additional permissions and the gateway is provisioned successfully.

image

**Comment by bmoore-msft** [comment]
closing as sample will need to be moved from the root
